common/variant/shredding.md

```diff
@@ -0,0 +1,244 @@
+| {a: [1,2,3], b: {c: null}} | null | null | untyped_value | untyped_value |
+| {a: 123, c: 456} | {c: 456} | null | typed_value | null |
+| {a: 123, b: {c: "hello", d: 456}} | null | {d: 456} | typed_value | typed_value |
+| [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | [{a: 1, b: {c: 2}}, {a: 3, b: {c: 4}}] | null | null | null |
+
+# Parquet Layout
+
+The **array** and **object** fields represent Variant array and object types, respectively. Arrays must use the three-level list structure described in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md.
+
+An **object** field must be a group. Each field name of this inner group corresponds to the Variant value's object field name. Each inner field's type is a recursively shredded variant value: that is, the fields of each object field must be one or more of **object**, **array**, **typed_value** or **untyped_value**.
+
+Similarly the elements of an **array** must be a group containing one or more of **object**, **array**, **typed_value** or **untyped_value**.
+
+Each leaf in the schema can store an arbitrary Variant value. It contains an **untyped_value** binary field and a **typed_value** field. If non-null, **untyped_value** represents the value stored as a Variant binary; the metadata and value of a normal Variant are concatenated. The **typed_value** field may be any type that has a corresponding Variant type. For each value in the data, at most one of the **typed_value** and **untyped_value** may be non-null. A writer may omit either field, which is equivalent to all rows being null.
+
+| typed_value | untyped_value | Meaning |
+|-------------|----------------|---------|
+| null | null | Field is missing in the reconstructed Variant. |
+| null | non-null | Field may be any type in the reconstructed Variant. |
+| non-null | null | Field has this column’s type in the reconstructed Variant. |
+| non-null | non-null | Invalid |
+
+The **typed_value** may be absent from the Parquet schema for any field, which is equivalent to its value being always null (in which case the shredded field is always stored as a Variant binary). By the same token, **untyped_value** may be absent, which is equivalent to their value being always null (in which case the field will always be missing or have the type of the **typed_value** column).
+
+The full metadata and value can be reconstructed from **untyped_value** by treating the leading bytes as metadata, and using the header, dictionary size and final dictionary offset to determine the start of the Variant value section. (See the metadata description in the common/variant/README.md for more detail on how to interpret it.) For example, in the binary below, there is a one-element dictionary, and the final offset (`offset[1]`) indicates that the last dictionary entry ends at the second byte. Therefore the full metadata size is six bytes, and the rest is the value section of the Variant.
```

Comment:
This is confusing to me. I thought that each variant value (top-level) had to share metadata and it was not allowed to have embedded metadata. Can individual variant values within a variant have a new metadata?

Reply:
The metadata is mainly the dictionary of all object keys. Once we shred, the keys that were associated with shredded fields are no longer needed (since they're stored as field names in the Parquet schema). The keys that are still needed (e.g. for an object that wasn't fully shredded) are stored as needed in the metadata of objects that use those keys. When reconstructing a full Variant value, the metadata will need to be rebuilt as the full value is built up from the shredded components.

The main motivation for not leaving a single metadata at the top level is that it would need to be fetched any time we fetch an `untyped_value` column. In a situation where some portion of the original value has very irregular keys from row to row, this could be a large penalty.

The two downsides I can see are:
1) Storing a small metadata next to every `untyped_value` adds overhead. Experimentally, this seems to be pretty small (~1%) after compression (I tested with snappy and zstd) if the metadata is regular.

Reply:
The metadata is mainly the dictionary of all object keys. Once we shred, the keys that were associated with shredded fields are no longer needed (since they're stored as field names in the Parquet schema). The keys that are still needed (e.g. for an object that wasn't fully shredded) are stored as needed in the metadata of objects that use those keys. When reconstructing a full Variant value, the metadata will need to be rebuilt as the full value is built up from the shredded components.

I'm open to other ideas, but the main motivation for not leaving a single metadata at the top level is that it would need to be fetched any time we fetch an `untyped_value` column. In a situation where some portion of the original value has very irregular keys from row to row, this could be a large penalty.

The downsides I can see are:

1) Needing to inspect and possibly rebuild each `untyped_value` when reconstructing the full Variant value. I think this is the biggest problem with this approach.
2) Storing a small metadata next to every `untyped_value` adds overhead. Experimentally, this seems to be pretty small (~1%) after compression (I tested with snappy and zstd) if the metadata is regular.
3) If the same key is used in different parts of the schema, they'll be duplicated. This doesn't seem likely to be a major issue, since the point of the metadata is to capture duplication from row to row, or in multiple array elements.

Reply:
This seems to add complexity to the format and cost. When writing, this must produce a filtered dictionary for each value. Those dictionaries may be overlapping and this could increase storage cost. And when reading the dictionaries would need to be merged. I think I prefer the simpler approach of keeping a single metadata/dictionary at the top level and reusing it to avoid all of the logic.

Reply:
I think the downsides are relatively small compared to the benefit on the read side can be significant in cases where only some fields are filtered or projected. It's not uncommon to have irregular metadata, making the metadata column quite expensive to fetch. If we need to fetch and decode the entire metadata to decode any untyped_value, it would create a very large penalty to read any scalar field that does not perfectly shred to its target type.